### ISSUE TITLE
docker_image broken on names.w.dots

### ISSUE THREAD

**Issue by nathwill** [bflad/chef-docker#356]
uh, no idea what's going on here, but i'm getting this error on both cases where we try to pull from a private repo; though the docker_image resource is working for the `centos` image?!

```ruby
           ================================================================================
           Error executing action `pull_if_missing` on resource 'docker_image[workspaces-web]'
           ================================================================================
           
           Docker::Error::ArgumentError
           ----------------------------
           Must have id, got: {"id"=>nil, :headers=>{}}
           
           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.0/lib/docker/base.rb:16:in `initialize'
           /tmp/kitchen/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.0/lib/docker/image.rb:117:in `new'
           /tmp/kitchen/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.0/lib/docker/image.rb:117:in `create'
           /tmp/kitchen/cache/cookbooks/docker/libraries/provider_docker_image.rb:55:in `pull_image'
           /tmp/kitchen/cache/cookbooks/docker/libraries/provider_docker_image.rb:105:in `block in <class:DockerImage>'
           /tmp/kitchen/cache/cookbooks/docker/libraries/provider_docker_image.rb:111:in `block in <class:DockerImage>'
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/treehouse-docker/recipes/workspaces.rb
           
            56:   docker_image "workspaces-#{img}" do
            57:     repo fqdi
            58:     tag 'production'
            59:   end
            60: end
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/treehouse-docker/recipes/workspaces.rb:56:in `block in from_file'
           
           docker_image("workspaces-web") do
             action [:pull_if_missing]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :docker_image
             cookbook_name "treehouse-docker"
             recipe_name "workspaces"
             repo "treehouse/workspaces.env.web"
             tag "production"
           end
```

we're definitely logged in and can successfully pull from the command-line :/. fwiw, we're using docker 1.7.1

**Comment by someara** [comment]
Hi. 

Sorry, this changed recently to match the docker remote-api The new syntax is:

docker_image 'fqdi/workspaces-web' 

or 

docker_image 'a chef resource name' do
  repo 'fqdi/workspaces-web'
  tag 'latest'
end

**Comment by nathwill** [comment]
hi @someara, yeah, the resource above is the version we'd changed to match the new resource, but we still got this problem.

the resource as described in the recipe looks like:

```ruby
docker_image 'workspaces-web' do
  repo 'treehouse/workspaces.env.web'
  tag 'production'
end
```

which seems to match what you described?

**Comment by nathwill** [comment]
maybe this is a bug in the docker-api gem, then?

**Comment by someara** [comment]
It's more likely a bug in my usage of it.

I don't think I tested names with dots in it... are those even allowed?

So far its been tested against all this:

https://github.com/bflad/chef-docker/blob/5883cbe3ffd680af637b04b88f2331d1f9b52038/test/cookbooks/docker_test/recipes/image_lwrp.rb

**Comment by nathwill** [comment]
it's definitely allowed; i'm pulling down a copy to test that recipe against and see if i can reproduce, will report back with any findings

**Comment by nathwill** [comment]
mmmmm yep; definitely something about the dot:

```ruby
       Recipe: docker_test::image_lwrp
        (up to date)
       
           
           ================================================================================
           Error executing action `pull_if_missing` on resource 'docker_image[Tom's container]'
           ================================================================================
           
           Docker::Error::ArgumentError
           ----------------------------
           Must have id, got: {"id"=>nil, :headers=>{}}
           
           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.0/lib/docker/base.rb:16:in `initialize'
           /tmp/kitchen/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.0/lib/docker/image.rb:117:in `new'
           /tmp/kitchen/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.0/lib/docker/image.rb:117:in `create'
           /tmp/kitchen/cache/cookbooks/docker/libraries/provider_docker_image.rb:59:in `pull_image'
           /tmp/kitchen/cache/cookbooks/docker/libraries/provider_docker_image.rb:109:in `block in <class:DockerImage>'
           /tmp/kitchen/cache/cookbooks/docker/libraries/provider_docker_image.rb:115:in `block in <class:DockerImage>'
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/docker_test/recipes/image_lwrp.rb
           
             9: docker_image "Tom's container" do
            10:   repo 'tduffield/testcontainerd.b'
            11: end
            12: 
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/docker_test/recipes/image_lwrp.rb:9:in `from_file'
           
           docker_image("Tom's container") do
             action [:pull_if_missing]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :docker_image
             cookbook_name "docker_test"
             recipe_name "image_lwrp"
             repo "tduffield/testcontainerd.b"
             tag "latest"
           end
```

i guess we can rename our images without too much trouble, but it's definitely weird, because we'd previously been running this with the shell-out version of the cookbook without bumping into this.

**Comment by someara** [comment]
Thanks!

For your reference, its consumed here:

https://github.com/bflad/chef-docker/blob/5883cbe3ffd680af637b04b88f2331d1f9b52038/libraries/provider_docker_image.rb#L58

... and ends up in an opts hash here

https://github.com/swipely/docker-api/blob/0735e4eb54176d0e39ef1eec129bbb91d65d579b/lib/docker/image.rb#L105

... which seems to be exactly what the docker cli was doing

https://github.com/docker/docker/blob/34a17a5de1d01a42e3e0173e5565b7de1af95bda/api/client/pull.go#L48-L51

**Comment by someara** [comment]
maybe it needs a shellwords escape or a %2E translation somewhere?

**Comment by nathwill** [comment]
yeah, i think something's going wrong in the assignment [here](https://github.com/swipely/docker-api/blob/0735e4eb54176d0e39ef1eec129bbb91d65d579b/lib/docker/image.rb#L117), but it's going to take me a while to debug, and at this point i'm pretty convinced that this is actually a bug in docker-api, even if it is needing to do the URI escaping (definitely a hunch worth tracking down). i don't want to waste your time on our edge-case if it's not a problem in your cookbook, so feel free to close this out and i'll pursue it with upstream once i nail it down. thanks for helping track it this far!

**Comment by nathwill** [comment]
by the way, love the cookbook improvements that have been landing lately, keep up the kickass work!

**Comment by someara** [comment]
thanks!
-s

**Comment by mrdziuban** [comment]
I'm running into this problem too. My image name doesn't have a period in it though, but it does have a hyphen, like `orgname/base-image`. My `docker_image` block looks like this:

```ruby
docker_image 'orgname/base-image' do
  repo 'orgname/base-image'
  tag 'latest'
  action :pull
end
```

@nathwill have you found any solutions for this?

**Comment by nathwill** [comment]
@mrdziuban not yet; for now we've reverted and pinned to docker cookbook version 0.37. it's on the TODO list, but... sooooo busy right now >__<

**Comment by mrdziuban** [comment]
Thanks, v0.37.0 worked for me as well! I'll stick to that for now.

**Comment by someara** [comment]
Hi. 

I can't seem to reproduce this username/name-with-dashes thing.

I've added dots and dashes to various image names in the test recipe here
https://github.com/bflad/chef-docker/commit/f43c12fe46ca5ac0db660d0aa53d4cf7fa8ed7b1

I can pull "hello-world', from the public hub, and both push-and-pull "someara/name-w-dashes".

While I was in there, I verified the name.w.dots fails for reference.

-s

**Comment by someara** [comment]
Actually, for name.w.dots, I'm getting a silent failure, not a Docker::Error::ArgumentError

**Comment by someara** [comment]
renaming this issue for my own sanity

**Comment by mrdziuban** [comment]
@someara I'm still having trouble pulling an image named `username/name-with-dashes` on v0.43.0. This is all that my cookbook is doing:

```
docker_service 'default' do
  action [:create, :start]
end

docker_registry 'https://index.docker.io/v1/' do
  username 'USERNAME'
  password 'PASSWORD'
  email 'EMAIL'
end

docker_image 'IMAGE_NAME' do
  repo 'IMAGE_NAME'
  tag 'latest'
  action :pull
end
```

I still get the error:

```
Must have id, got: {"id"=>nil, :headers=>{}}
```

**Comment by mrdziuban** [comment]
Update on my situation:

It seems that it's not related to the dashes in the name. I think my issue actually has to do with slashes in the name, i.e. pulling from an organization's repo. I created [this public test repo](https://hub.docker.com/r/mrdziuban/testrepo/) to demonstrate the issue. Try using this to pull the image:

```
docker_image 'mrdziuban/testrepo' do
  repo 'mrdziuban/testrepo'
  tag 'latest'
  action :pull
end
```

**Comment by dimfeld** [comment]
I've just encountered the same thing, but in my case the workaround was to remove the `tag` from the resource. No dots here: the image in this case is `carevoyance/qubit-bamboo` and the tag is simply `latest`.  

Here's the Chef block I'm using, nothing unusual:

```
docker_image node[:bamboo][:docker_image] do
    action :pull
    tag node[:bamboo][:version]   # Works only when this is commented out.
end
```

And the compiled version:

```
 docker_image("carevoyance/qubit-bamboo") do
   action [:pull]
   retries 0
   retry_delay 2
   default_guard_interpreter :default
   declared_type :docker_image
   cookbook_name "cv_mesos"
   recipe_name "qubit-bamboo"
   tag "latest"
   api_retries 3
   repo "carevoyance/qubit-bamboo"
 end
```

**Comment by psi** [comment]
I'm seeing the same behavior with the `amazon/amazon-ecs-agent` image on Docker Hub.

ubuntu 14.04
chef-client 12.4.1
docker cookbook 1.0.18

```
  * docker_image[amazon/amazon-ecs-agent] action pull

    ================================================================================
    Error executing action `pull` on resource 'docker_image[amazon/amazon-ecs-agent]'
    ================================================================================

    Docker::Error::ArgumentError
    ----------------------------
    Must have id, got: {"id"=>nil, :headers=>{}}

    Cookbook Trace:
    ---------------
    /var/chef/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/base.rb:16:in `initialize'
    /var/chef/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/image.rb:117:in `new'
    /var/chef/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/image.rb:117:in `create'
    /var/chef/cache/cookbooks/docker/libraries/provider_docker_image.rb:75:in `pull_image'
    /var/chef/cache/cookbooks/docker/libraries/provider_docker_image.rb:137:in `block in <class:DockerImage>'

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/r5-ecs-host/recipes/ecs_agent.rb

     25: docker_image 'amazon/amazon-ecs-agent' do
     26:   action :pull
     27:   notifies :redeploy, 'docker_container[ecs-agent]'
     28: end
     29:

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/r5-ecs-host/recipes/ecs_agent.rb:25:in `from_file'

    docker_image("amazon/amazon-ecs-agent") do
      action [:pull]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :docker_image
      cookbook_name "r5-ecs-host"
      recipe_name "ecs_agent"
      tag "latest"
      api_retries 3
      repo "amazon/amazon-ecs-agent"
    end
```

From my recipe:

```ruby
# I've tried numerous variations on this using both 'latest' and 'v1.4.0'
# for the tag and including or not including the repo attribute. All behave the same.
docker_image 'amazon/amazon-ecs-agent' do
  action :pull
  notifies :redeploy, 'docker_container[ecs-agent]'
end

docker_container 'ecs-agent' do
  image 'amazon/amazon-ecs-agent'
  env %W(
    ECS_LOGFILE=/log/ecs-agent.log
    ECS_DATADIR=/data
    ECS_CLUSTER=#{node['amazon-ecs-agent']['cluster']}
    ECS_ENGINE_AUTH_TYPE=#{node['amazon-ecs-agent']['engine_auth_type']}
    ECS_ENGINE_AUTH_DATA=#{node['amazon-ecs-agent']['engine_auth_data'].to_json}
  )
  volumes([
    "/var/run/docker.sock:/var/run/docker.sock",
    "#{node['amazon-ecs-agent']['log_dir']}:/log",
    "#{node['amazon-ecs-agent']['data_dir']}:/data"
  ])
  port '51678:51678'
  detach true
end
```

**Comment by psi** [comment]
Hmmm, ok... So, in my case, adding `tag` and `repo` attributes works if I'm not otherwise authenticated to a private registry.

**Comment by someara** [comment]
@dimfeld Can you test if you're running into #379 ?

**Comment by dimfeld** [comment]
#379 only applies if you've authenticated to a private registry, right? In my case I do have a private registry set up, but there's no authentication involved. (That is, I'm not using the `docker_registry` resource or anything similar.) 

I did try removing all recipes that use the private registry from the runlist and it didn't make a difference. Is that sufficient to rule it out?

**Comment by someara** [comment]
Well... it's definitely a clue. Are you pulling from public, then private? Or private, then public? Or..

**Comment by dimfeld** [comment]
It was private, then public. The problem occurred in odd clusters. It would fail for a while, and then start working again, and then fail again, and now it works, with no changes in any of the docker-related stuff. I was messing around with some DNS configuration but the cookbook should have failed well before it got to the docker stuff if that wasn't configured right.

Unfortunately, I'm not able to get it to reproduce tonight, so I'm not sure if I can do anything else to help for now, but I'll chime back in if I see this again.

**Comment by someara** [comment]
This should be fixed in v1.0.21... patched the docker-api gem via #379
